gadgets/trace_capabilities/program.bpf.c

```diff
@@ -0,0 +1,334 @@
+		// override_creds/revert_creds (e.g. during overlayfs cache or copyup)
+		// https://kernel.org/doc/html/v6.2-rc8/security/credentials.html#overriding-the-vfs-s-use-of-credentials
+		return 0;
+	}
+
+	pid_tgid = bpf_get_current_pid_tgid();
+	pid = pid_tgid >> 32;
+
+	if (pid == my_pid)
+		return 0;
+
+	if (targ_pid != -1 && targ_pid != pid)
+		return 0;
+
+	if (audit_only) {
+		if (LINUX_KERNEL_VERSION >= KERNEL_VERSION(5, 1, 0)) {
+			if (cap_opt & CAP_OPT_NOAUDIT)
+				return 0;
+		} else {
+			if (!cap_opt)
+				return 0;
+		}
+	}
+
+	if (unique) {
+		struct unique_key key = {
+			.cap = cap,
+			.mntns_id = mntns_id,
+		};
+
+		if (bpf_map_lookup_elem(&seen, &key) != NULL) {
+			return 0;
+		}
+		u64 zero = 0;
+		bpf_map_update_elem(&seen, &key, &zero, 0);
+	}
+
+	struct args_t args = {};
+	args.current_userns =
+		(u64)BPF_CORE_READ(task, real_cred, user_ns, ns.inum);
+	args.target_userns = (u64)BPF_CORE_READ(targ_ns, ns.inum);
+	/*
+	 * cap_effective has kernel_cap_t for type.
+	 * This type definition changed along the time:
+	 * 1. It was defined as a __u32 in:
+	 * https://github.com/torvalds/linux/commit/1da177e4c3f4
+	 * 2. It later was modified to be an array of __u32, so 64 bits kernel
+	 * can use 64 bits for capabilities while supporting legacy 32 bits
+	 * ones:
+	 * https://github.com/torvalds/linux/commit/e338d263a76a
+	 * 3. It was recently defined to be a simple u64:
+	 * https://github.com/torvalds/linux/commit/f122a08b197d
+	 * BPF_CORE_READ_INTO() will handle the different size for us and in any
+	 * case, we define args.cap_effective as u64 which is enough to contain
+	 * the information.
+	 */
+	BPF_CORE_READ_INTO(&args.cap_effective, task, real_cred, cap_effective);
+	args.cap = cap;
+	args.cap_opt = cap_opt;
+	bpf_map_update_elem(&start, &pid_tgid, &args, 0);
+
+	return 0;
+}
+
+SEC("kretprobe/cap_capable")
+int BPF_KRETPROBE(ig_trace_cap_x)
+{
+	__u64 pid_tgid;
+	__u64 uid_gid = bpf_get_current_uid_gid();
+	struct args_t *ap;
+	struct cap_event *event;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+	ap = bpf_map_lookup_elem(&start, &pid_tgid);
+	if (!ap)
+		return 0; /* missed entry */
+
+	event = gadget_reserve_buf(&events, sizeof(struct cap_event));
+	if (!event)
+		return 0;
+
+	event->current_userns = ap->current_userns;
+	event->target_userns = ap->target_userns;
+	event->cap_effective = ap->cap_effective;
+	event->pid = pid_tgid >> 32;
+	event->tgid = pid_tgid;
+	event->cap = ap->cap;
+	event->uid = (u32)uid_gid;
+	event->gid = (u32)(uid_gid >> 32);
+	event->mntnsid = gadget_get_mntns_id();
+	bpf_get_current_comm(&event->task, sizeof(event->task));
+	event->ret = PT_REGS_RC(ctx);
+	event->kstack = gadget_get_kernel_stack(ctx);
+	event->timestamp = bpf_ktime_get_boot_ns();
+
+	if (LINUX_KERNEL_VERSION >= KERNEL_VERSION(5, 1, 0)) {
+		event->audit = (ap->cap_opt & CAP_OPT_NOAUDIT) == 0;
+		event->insetid = (ap->cap_opt & CAP_OPT_INSETID) != 0;
+	} else {
+		event->audit = ap->cap_opt;
+		event->insetid = -1;
```

Comment:
I don't recommend to use `bpf_core_type_matches` here, because `struct cred` also changed in Linux 6.7-rc6 [1].

Because the `vmlinux.h` in IG is version `6.6`, and my local environment is `6.10`, I spent a lot of time looking for why the `bpf_core_type_matches` returns false. In such cases, seems we need to trace all the nested structures, and provide headers for each version. What do you think?

[1] https://github.com/torvalds/linux/commit/f8fa5d76925991976b3e7076f9d1052515ec1fca

Reply:
I guess your motivation for using `bpf_core_type_matches` is that some distributions may pick patches or do certain backports, making it better to judge the structure than the version number. 

However, we cannot predict which of the two patches (1: updating `struct cred`, 2: updating `union security_list_options`) is included in the user's Linux distribution. This leads to four scenarios, with potentially more combinations in the future.